Allow midi CC channel filtering

bornej opened the pull request:
Hi, 
This PR add [midi:ctrl num chan] metadata for UI elements and allow
MIDI CC channel filtering: UI Elements can now receive and send CC value from/to
a specified MIDI channel.

I tested, documented my modifications.
This is my first attempt contributing to a project,
I Hope I am doing this right...

Best regards...

Rem: most MIDI hardware use channel values from 1 to 16.
So [midi:ctrl num 0] will mean sending/receiving midi CC to the synth/controller... channel 1.
This look "unnatural" however when monitoring midi data coming from my synths 
with `aseqdump -p` I observed that:

synth Ch 1 -> aseqdump Ch 0
synth Ch 2 -> aseqdump Ch 1
...
synth Ch 16 -> aseqdump Ch 15

So I decided to keep the channel range to [0,15].

sletz commented:
Thanks for the PR proposal. If you go this direction (adding a chan field), then it would be better to provide a complete implementation:

- move `fChanFilter` of the base uiMidi class, rename it `fChan`
- implement channel support in all supported types
- update the DSP tests and documentation accordingly

bornej commented:
Hi,
That's ok for me. I'll be working on this.
If everything turns ok the complete implementation/tests/documentation should be here soon.

Thanks for your answer!

bornej commented:
Hi,
Midi channel filtering is now implemented for the following types:
`ctrl`, `keyon`, `keyoff`, `keypress`, `pitchbend`, `prgChange`.

I added documentation and tests in `midiTester.dsp`.

About the other types:
- `start`, `stop`, `clock` -> don't have channel information/ don't need channel filtering
- `chanpress` -> Not documented/not sure what this does...

Also while working on this I thought about some new propositions:
- **Program Changes**: I think we should remove the num field in 
[midi:pgm num] and [Midi:pgm num chan]. It seems strange 
to create an UI element that listen to only one program number.
We should probably listen to any incoming program change value...

- **Pitchbend**: As a user I would rather have the UI component
give me the actual/raw pitchbend value in range {-8191,8192} and
re-normalize it myself, so I can chose freely what to do with those values. 
While I understand the idea of the normalization added in
`commit f60e87f304353d810c344f247190fb446dd42e33` I think it
would be more coherent to keep the input values untouched... 
I'm saying this because every other UI component just listen to the input values and spit them out, so I was expecting the same behavior for the pitchbend and couldn't get my test programs working. (documentation doesn't mention the normalization yet).
When I reverted the changes from the commit I could work with pitchbend again with ease... It is also possible that normalization here serve another purpose that
I've missed... 

I would be happy to continue working on this if it helps. 
Again I hope, i'm doing this right...

Best regards.

sletz commented:
Thanks for the proposal. 
I finally reworked your PR a bit and merged it in https://github.com/grame-cncm/faust/commit/d68efbc1c67aec84489e456a06cf68c073ebba06

bornej commented:
Thank you very much, this look nice to me.
 
In midi.md documentation could you just change this 

```
Note that all handlers have a version with an explicit MIDI channel in range {0, 15},
and a version without any MIDI channel (so receiving/sending from/to any MIDI 
channel).
```

to 

```
Note that all handlers have a version with an explicit MIDI channel in range {0, 15},
and a version without any MIDI channel (so receiving from any MIDI channel and sending back to channel 0).
```
?
I realized that the handlers without midi channel always send their output values to channel 0, independently of the input channel... I don't think this is problematic as there
is now an explicit way to filter channels...

Otherwise, I ran tests again and everything related to midi channels seems to work.
So I feel you can close this PR.

Thanks you again!
Hope this helped.

Best regards,
Jonathan.

PS: The behaviour of the pitchwheel since commit f60e87f304353d810c344f247190fb446dd42e33
wasn't working well for me. I couldn't get the full range of my keyboards pitchbend controllers... Reversing the commit solved my problem. I guess this is the subject for another discussion but I had this small issue while working on this PR...

sletz commented:
- thanks for the documentation improvement 
- and yes the pitchbend behavior is a bit unorthodox, we adopted this way at some point and I can not remember the exact reason.... possibly @rmichon remembers better..
- BTW I'm curious... what kind of sophisticated MIDI controllable DSP are you implementing ?

bornej commented:
> thanks for the documentation improvement
I don't want to be blamed for misleading faust users so I
try to take documentation seriously ^^.

> and yes the pitchbend behavior is a bit unorthodox, we adopted this way at some point > and I can not remember the exact reason.... possibly @rmichon remembers better..
I will try to ask him then.

> BTW I'm curious... what kind of sophisticated MIDI controllable DSP are you 
> implementing ?
I'm working on Mass-interraction models built with mi_faust.
For the moment I try to create percussive instruments to play/tune with my Clavia Nord-Drum.
I only had time to experiment with simple models, so it's not that sophisticated yet,
but what I heard sounds really promising!
I'm really looking forward to bringing new development in Mass-interraction modeling for faust. I'm in contact with the @mi-creative team and we are supposed to 
meet with @rmichon soon on this subject... Looking forward to that too...

Also, If you are interested I plan to make some audio/video demos... I hope this
will be coming soon. Don't hesitate to contact me by mail (bornej@pm.me)
as it might also be simpler in french...

Thanks again for everything! 

PS:
In the end channel filtering wasn't needed for this project.
but well... I'm sure it would have been needed sometime, somewhere by someone  anyway...

bornej commented:
Please wait before closing, there might be a problem with midi notes...

bornej commented:
Okay, seems like I read your commit a bit too fast and my tests didn't cover 
the midi:key handler...
There at least this to fix:
```
diff --git a/architecture/faust/gui/MidiUI.h b/architecture/faust/gui/MidiUI.h
index c30cba51b..87a156405 100644
--- a/architecture/faust/gui/MidiUI.h
+++ b/architecture/faust/gui/MidiUI.h
@@ -588,7 +588,7 @@ class MidiUI : public GUI, public midi
                             fKeyOffTable[num].push_back(new uiMidiKeyOff(fMidiHandler, num, this, zone, min, max, input, chan));
                         } else if (gsscanf(fMetaAux[i].second.c_str(), "keyoff %u", &num) == 1) {
                             fKeyOffTable[num].push_back(new uiMidiKeyOff(fMidiHandler, num, this, zone, min, max, input));
-                        } else if (gsscanf(fMetaAux[i].second.c_str(), "key %u %u", &num, &chan) == 1) {
+                        } else if (gsscanf(fMetaAux[i].second.c_str(), "key %u %u", &num, &chan) == 2) {
                             fKeyTable[num].push_back(new uiMidiKeyOn(fMidiHandler, num, this, zone, min, max, input, chan));
                         } else if (gsscanf(fMetaAux[i].second.c_str(), "key %u", &num) == 1) {
                             fKeyTable[num].push_back(new uiMidiKeyOn(fMidiHandler, num, this, zone, min, max, input));
@@ -701,9 +701,9 @@ class MidiUI : public GUI, public midi
                     int channel_aux = fKeyTable[note][i]->fChan;
                     if (channel_aux == -1 || channel == channel_aux) {
                         if (fTimeStamp) {
-                            fKeyTable[note][i]->modifyZone(date, 0);
+                            fKeyTable[note][i]->modifyZone(date, FAUSTFLOAT(velocity));
                         } else {
-                            fKeyTable[note][i]->modifyZone(0);
+                            fKeyTable[note][i]->modifyZone(FAUSTFLOAT(velocity));
                         }
                     }
                 }

```
Also now that I'm seeing that,  I feel like I don't really grasp how the midi:key 
works...

Will need some time to investigate more and run some more tests...
I have trust issues about this now...

bornej commented:
Things seems to be back to
normal... I should have caught that mistake because the documentation example for midi:key clearly wasn't working (even before your modifications)...
This means I wrote this test and forgot to even run it. Really sloppy from my part. And I hate sloppyness. So, really sorry about that.

I'm still taking time to re-read how midi:key is handled...

sletz commented:
So do you have a patch?

bornej commented:
I put a diff in the comment above... Is it ok?

sletz commented:
OK, done in https://github.com/grame-cncm/faust/commit/ae8a7620da41292fae0fe8fa2b43a6aa45cc1eaa

bornej commented:
Testing (everything this time)...

bornej commented:
- Ok: ctrl, keyon, keyoff, key, clock (with and without channel filtering)
- Couldn't test: keypress(aftertouch), start, stop.
- Still couldn't get pitchwheel working: (I would suggest reversing f60e87f304353d810c344f247190fb446dd42e33)
- Found a small typo on documentation (see next comment for patch).

Appart from the 3 missing tests (don't have aftertouch and start stop on my controllers)
This is starting to look good...

bornej commented:
diff --git a/documentation/faust-manual/src/midi.md b/documentation/faust-manual/src/midi.md
index cf76807ca..14ccd676e 100644
--- a/documentation/faust-manual/src/midi.md
+++ b/documentation/faust-manual/src/midi.md
@@ -207,7 +207,7 @@ released.
 <!-- faust-run -->
 ```
 import("stdfaust.lib");
-freq = hslider("frequency[midi:keyon 62]",200,50,1000,0.01) : si.smoo;
+freq = hslider("frequency[midi:keyoff 62]",200,50,1000,0.01) : si.smoo;
 process = os.sawtooth(freq);
 ```
 <!-- /faust-run -->
@@ -244,7 +244,7 @@ released.)
 <!-- faust-run -->
 ```
 import("stdfaust.lib");
-freq = hslider("frequency[midi:keyon 62 3]",200,50,1000,0.01) : si.smoo;
+freq = hslider("frequency[midi:keyoff 62 3]",200,50,1000,0.01) : si.smoo;
 process = os.sawtooth(freq);
 ```
 <!-- /faust-run -->

sletz commented:
Thanks, commited in https://github.com/grame-cncm/faust/commit/2bbaef1916f77ee48f44b67033855fec60a0390d
We will not revert f60e87f for now. This behavior had a good reason to be, I have to check with @rmichon again.

sletz commented:
Pitch wheel handing has been completely reworked in https://github.com/grame-cncm/faust/commit/3ca5f23355a635251b0cf0ea2a58b0ec528552ac. Basically:
- the `hslider("bend" [midi:pitchwheel] default min max step) `works as other ones.
- the MidiUI stuff does a 2 zones linear mapping [0 8191] <==> [min 0] and mapping [8191 16383] <==> [0 max]
- new `ba.semi2ratio` and `ba.ratio2semi`  functions (https://github.com/grame-cncm/faustlibraries/blob/master/basics.lib) have to be used to convert semitones into a frequency ratio
- RPN0 handing has been added to possibly change the pitch wheel [min, max] range while the application is running

sletz commented:
Can you possibly test the new implementation (so that we can possibly close this issue) ? Thanks.

bornej commented:
Yes, by the way thanks for your modifications.
I'm taking some time today to read them and test the new functionalities...
Coming back to you soon.
